### DISCUSSION TITLE
Unexepected implies warning: cannot resolve warning and unexpected solution

### DISCUSSION THREAD

**Discussion by mrkento11** [yalmip/YALMIP#1221] [Problems using YALMIP] [answered]
Dear Forum,
I have been trying to solve a Power System Generation Expansion Optimization problem. Given that it is a loop that solves the mathematical problem, and then in a second phase executes some tests as a verifier, I built an optimizer object to repeatedly solve the similar problem. I want to constrain the addition of both solar and wind as to make additions in a not so disproportionate way, thus I want to limit that for any given year in the future, the total energy of any lump technology cannot be more than 3 times the amount of the other. I tried to resolve this by adding an implies with the following code:
```
d = binvar(2, study_horizon,'full'); % The study horizon is 30 years, so this variable is a matrix 2x30.
...
for t = 1:study_horizon      % the array VarConstraints is built only for the implies. When calling the optimizer command I call it like: optimizer(Constraints + VarConstraints,Cost,options,{...},{...})
   VarConstraints = VarConstraints + [sum(d(:,t)) == 1,    % This line says that the solver must choose one of the following cases
                               implies(d(1,t),[0 <= sum(Exist_git(solar,t)) + sum(Cand_git(solar,t)) <= 3*(sum(Exist_git(wind,t)) + sum(Cand_git(wind,t)))]);       % This lines can be read as: 0 <= sum(Energy from solar at year t) <= 3*sum(Energy from wind at year t)
                               implies(d(1,t),[0 <= sum(Exist_git(wind,t)) + sum(Cand_git(wind,t)) <= 3*(sum(Exist_git(solar,t)) + sum(Cand_git(solar,t)))])];      % This lines can be read as: 0 <= sum(Energy from wind at year t) <= 3*sum(Energy from solar at year t)
end
```

I am trying to add some bounds because I am seeing a Warning that reads: "_Warning: You have unbounded variables in an implication leading to a lousy big-M relaxation._ "

I am seeing that the decision variables involved are Exist_git and Cand_git, but I have a previous constraint over Exist_git and Cand_git that read:
```
for t = 1:study_horizon
    Constraints = Constraints + [(0 <= Exist_git(:,t) <= Exist_Upper_Limit): 'Existing Generator Bounds'];  % There is no decision regarding existing infrastructure
    Constraints = Constraints + [(0 <=  Cand_git(:,t) <=xit(:,t).*Candidate_Upper_Limit): 'Candidate Generator Bounds']; 
end
```
Xit is the integer decision variable about constructing or not technology i at year t. In case it decides to invest in that technology, it has the option to make integer amount of investments across time. 
Now, we observe that one is explicitly bounded, while the other is only implicitly bounded. Nonetheless, even when we see the warning we are observing a good or plausible answer as reads:
![image](https://user-images.githubusercontent.com/111488263/222396345-ac92ebe3-2113-40c3-9ef3-134199ad415d.png)
 As we can see in the image the quotient of Solar-Cand and Wind-Cand is about 3:1. So the solver decides to invests in 3 times many solar generators as wind. 

However, to eliminate the Warning I add an explicit (trivial) bound over the decision variable Cand_git that reads:
```
for t = 1:study_horizon
     VarConstraints = VarConstraints + [0 <= Cand_git(:,t) <= Max_Install]; % 
end 
```
MaxInstall is an array that dictates the maximum amount we can install for a given technology. For instance, solar and wind have a top maximum total amount of 40,000 MW. 
Moreover, I have some other constraints relating to this topic that read:

```
for t = 1:study_horizon
    Constraints = Constraints + [xit(:,t).*Cand_Upper <= Max_Install] % xit is the integer decision variable decision v
end
```
To be clear, again, xit is the integer decision variable to invest or not in any given technology. So, Cand_Upper is the Installed Capacity of a single unit of technology i. For example, I am modelling solar with an individual upper limit of 2,250 MW and with a Max Install total of 40,000MW. Other technologies have different parameters.

Coming back to the original point, when I add the trivial constraint over Cand_git, the implies warning disappears, but the solution is unexpected as shown in the following figure:
![image](https://user-images.githubusercontent.com/111488263/222400522-6aaf0573-dba6-4a8f-9c80-74aedbc2e79e.png)

As you can see in year 2029, as well, in previous years, the quotient of Solar-Cand and Wind-Cand is about 7:1. Moreover, this is a very unexepected result as the added constraint was nothing but a triviality implied within other constraints. But it appears to not be respecting the implies constraints (VarConstraints)

Is there something I am misinterpreting? Is there something flawed within my code? I can share the code, but it is astoundingly long, although easy to follow as it is heavily commented.

Finally, I would like to ask what to do to be able to participate in the Google Forums group, as I initially intended to make this query there, but I learnt that I cannot open a new discussion there.

**Comment by johanlofberg** [comment, marked as answer]
Don't remove the explicit bound on Cand_git, it just leads to numerically bad models

You never use d(2,t) which is weird

**Reply by mrkento11** [reply]
Dear Professor Löfberg,
Thanks as usual for your speedy answer.

You are right that when I wrote this Forum Question, I didn't use d(2,t). However, it was only a mistake when writing in the discussion. My code correctly was using the alternative of the implies:

![image](https://user-images.githubusercontent.com/111488263/222661788-437c5b01-945e-4f23-acac-7bb124a041ac.png)

As well, you can see in the previous screenshot that I am considering the explicit bound on Cand_git. However, I am experiencing unexpected results as shown below:
![image](https://user-images.githubusercontent.com/111488263/222664059-9836f698-94d4-4bf5-bb4c-b7b21025e06b.png)

I believe that when I consider the explicit bounds, the implies related constraints (VarConstraints) are not being respected. In the sense that the solver decides to not even decide for a single Wind Candidate power plant, thus the energy that comes from solar is more than 3 times the energy that comes from wind.

Do you have any idea what I might be doing wrong? Is there something wrong with the logic? Is there something wrong with the translation of this  mathematical formulation?

**Reply by johanlofberg** [reply]
You would have to supply reproducible example. Impossible to say anything from partial screen shots

google groups is admined by google, I do not control that in any sense so how to join or why one cannot join is not something I know anything about

**Reply by mrkento11** [reply]
Thank you Professor Löfberg,
Just before making the attempt to supply a reproducible example, I just have the trivial question that when using an implies() when creating an optimizer object, do we also need to bound the Parameters?
My question is based on the fact that when declaring optimizer Parameters, they are declared as variables. And only are differentiated when we call the function optimizer().
Thus, again, my question would be if when defining and collecting the constraints, should we also trivially and explicitly bound the Parameters in our optimizer objects.

Thanks.

**Reply by johanlofberg** [reply]
Yes, during modelling YALMIP makes no difference between parameters and decision variables, hence all parameters must have bounds also for the big-M modelling to work

**Reply by mrkento11** [reply]
Thanks again for your clarifications Professor Löfberg.

Appended you can find a much reduced (but still complete) formulation of my Generation Expansion Problem solved using Yalmip + Gurobi. 
As you can observe, in case I don't make explicit bounds for Cand_git (lines 167-172) Yalmip warns me of a lousy big-M relaxation but gives me expected results relating to investments in Solar and Wind. However, if I do in fact state the explicit bounds, the warning goes away, but I interpret the constraints in the implies are not abided.

Thanks for your help.

The zip file contains a Live Script file and a mat file. The Live Script file is a 20-line (plus two somewhat long functions), while the mat file just holds a few variables to make the script work.

[test_yalmip_github.zip](https://github.com/yalmip/YALMIP/files/10915802/test_yalmip_github.zip)

**Reply by johanlofberg** [reply]
My guess is that you are running into scaling problem. The big-modelling engine assumes well-scaled problems which means unbounded variables uses upper bounds of 10^4. You can test this by editing derivebounds.m and change the constant 10^4 to something bigger (the price you pay is that the model will be horrible numerically)

**Reply by johanlofberg** [reply]
Also, it is always recommended to first debug without using optimizer, but instead as in this case by setting up the model and then solving with the constraint parameter = values added to the model. That way you remove the possibility that it is the optimizer which has bugs

**Reply by johanlofberg** [reply]
Also, when supplying examples to debug, you must remove everything not necessary for the bug to occur. As you only talk about wind and solar, remove everything else (while still seeing the thing you don't expect). Also, if it happens on one year, why include other years. For an outsider, this code is simply to complex to understand, you should be able to distill it significantly down to a minimal reproducible example. Remove simplify simplify

**Reply by johanlofberg** [reply]
If your issues is that the "Solar-cand" row should not be larger than 3x"wind-cand", it is not due to any bug in implies, as manual modelling yields the same


```
        aa = sum(Exist_git(Flags.flag_EcoThermal_exist+1:Flags.flag_EcoThermal_exist+first_wind_exist-1,:,t)*(8760*GEP_Load.LD)) + sum(Cand_git(Flags.flag_EcoThermal_cand+1:Flags.flag_EcoThermal_cand+first_wind_cand-1,:,t)*(8760*GEP_Load.LD)) ;
        bb = 3*(sum(Exist_git(Flags.flag_EcoThermal_exist+first_wind_exist:Flags.flag_renewables_exist,:,t)*(8760*GEP_Load.LD)) + sum(Cand_git(Flags.flag_EcoThermal_cand+first_wind_cand:Flags.flag_renewables_cand,:,t)*(8760*GEP_Load.LD)));
        cc = sum(Exist_git(Flags.flag_EcoThermal_exist+first_wind_exist:Flags.flag_renewables_exist,:,t)*(8760*GEP_Load.LD)) + sum(Cand_git(Flags.flag_EcoThermal_cand+first_wind_cand:Flags.flag_renewables_cand,:,t)*(8760*GEP_Load.LD));
        dd = 3*(sum(Exist_git(Flags.flag_EcoThermal_exist+1:Flags.flag_EcoThermal_exist+first_wind_exist-1,:,t)*(8760*GEP_Load.LD)) + sum(Cand_git(Flags.flag_EcoThermal_cand+1:Flags.flag_EcoThermal_cand+first_wind_cand-1,:,t)*(8760*GEP_Load.LD)));
        VarConstraints = VarConstraints +  [sum(d(:,t)) == 1;       
                    %    implies(d(1,t),[0 <= aa]);
                     %   implies(d(1,t),[aa <= bb]);
                      %%  implies(d(2,t),[0 <= cc])
                        implies(d(2,t),[cc <= dd])
                     aa >= -1e12*(1-d(1,t));
                     bb - aa >= -1e12*(1-d(1,t));
                     cc >= -1e12*(1-d(2,t));
                     dd - cc >= -1e12*(1-d(2,t));
                        ];
```

**Reply by mrkento11** [reply]
Thank you for your continuous help Professor.

I will go back to checking the mathematical formulation to check my mistake, and if I indeed translated correctly to Yalmip.

Moreover, I apologize for not simplifying enough. Will try to do a better job at simplification in case of future questions.

Thanks.

**Reply by johanlofberg** [reply]
Easiest is to skip optimizer construction and make call directly using optimize with added constraints on parameters, as you then directly can see if any constraint is violated (i.e. is there a bug in solver), and check the values of variables and constraint that you think are wrong. Using optimizer is hard to debug as it doesn't assign any values so you cannot check things easily

**Reply by mrkento11** [reply]
Just to let you know Professor, I found out the mistake and it was a reasoning mistake rather than a coding or software mistake (Yalmip).
I was experiencing unexpected results because as you point out I was seeing that "Solar-cand" should not be larger than 3*"wind-cand". I wanted to solar not grow disproportionately bigger than wind, so I tried to add a constraint for that. In the small test case I am using solar is much more attractive than wind, but I wanted to also consider if wind might be more attractive than solar. I thought that a good approach was to add a constraint for solar <= 3xWind, but also add in case wind is more attractive to be wind <= 3xSolar. {OR(solar <= 3xWind, wind <= 3xSolar)}.

I was using implies somewhere else within the whole code and thought that I should use it there as well. But, as you pointed out the manual modeling gave out the same answer. I observed that implies (obviously) decides to abide by the more lenient constraint among the two, and the constraint {wind <= 3*solar} was always being abided. Therefore, I realized that it was a reasoning mistake on my behalf to think that I need an implies constraint there (no implies needed), when I should add both constraints, and let the solver find out which of them both is the restricting constraint {AND(solar <= 3xWind, wind <= 3xSolar)}.

Thank you for your help on this topic, Professor.
